### ISSUE TITLE
Count() does not equal number of results if included navigation property does not match query filter

### ISSUE THREAD

**Issue by tmthill** [dotnet/efcore#19649]
If a query selects a property from a related entity reached by a required navigation property, but the related entity is excluded by a query filter, `query.Count()` will not match `query.ToList().Count`. Actualizing the results with `ToList()` will exclude the row from its results, as the related entity is filtered out, but `Count()` does not exclude the result even though the query is selecting a member of the navigation property explicitly.

We use this functionality to build search results one page at a time: build a filtered and projected `IQueryable<TEntity> query`, use `query.Count()` to get the total number of results, then `query.OrderBy(orderBy).Skip(skip).Take(take).ToList()` to resolve a subset of the results. Our actual implementation has a much more complicated projection thanks to AutoMapper, but as recently as EF Core 2.2.6, calling `Count()` on the queryable returned the same cardinality as the result set (indeed, the code generated for SQL Server has the same FROM and JOIN clauses and differs only in the SELECT lists).

### Steps to reproduce

``` C#
using Microsoft.EntityFrameworkCore;
using System.Linq;
using Xunit;

namespace EntityFrameworkCoreTests
{
    class Entity
    {
        public int EntityId { get; set; }
        public int RelatedEntityId { get; set; }
        public virtual RelatedEntity RelatedEntity { get; set; }
    }

    class RelatedEntity
    {
        public int RelatedEntityId { get; set; }
        public bool IsDeleted { get; set; }
        public string SomeProperty { get; set; }
    }

    class TestContext : DbContext
    {
        public TestContext(DbContextOptions<TestContext> options) : base(options) { }

        public DbSet<Entity> Entities { get; set; }
        public DbSet<RelatedEntity> RelatedEntities { get; set; }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<RelatedEntity>()
                .HasQueryFilter(relatedEntity => !relatedEntity.IsDeleted);
        }
    }

    public class CountDoesNotMatchTests
    {
        [Fact]
        public void Queryable_Count_should_match_ToList_Count()
        {
            using var context = new TestContext(
                new DbContextOptionsBuilder<TestContext>()
                    .UseInMemoryDatabase(GetType().Name)
                    .Options);

            var entity = new Entity
            {
                RelatedEntity = new RelatedEntity
                {
                    IsDeleted = true
                }
            };

            context.Add(entity);
            context.SaveChanges();

            var query = context.Entities
                .AsNoTracking()
                .Select(e => e.RelatedEntity.SomeProperty);

            Assert.Equal(query.ToList().Count, query.Count());
            // query.ToList().Count is zero; query.Count() is one
            // As of EF Core 2.2.6, they were both zero
        }
    }
}
```

The code above uses InMemory for brevity, but Sqlite exhibits the same behavior and generates the following query for `query.ToList()`, returning no rows:
``` SQL
SELECT "t"."SomeProperty"
FROM "Entities" AS "e"
INNER JOIN (
    SELECT "r"."RelatedEntityId", "r"."IsDeleted", "r"."SomeProperty"
    FROM "RelatedEntities" AS "r"
    WHERE NOT ("r"."IsDeleted")
) AS "t" ON "e"."RelatedEntityId" = "t"."RelatedEntityId"
```

And for `query.Count()`, returning a count of 1:
``` SQL
SELECT COUNT(*)
FROM "Entities" AS "e"
```

I also noticed that, if I used `.Include(e => e.RelatedEntity)` instead of `.Select(e => e.RelatedEntity.SomeProperty)` in the test above, EF Core 2.2.6 and 3.1.1 both return 1 for `query.Count()` and 0 for `query.ToList().Count`. This is not currently causing us issues, but it looks similar yet behaves consistently across versions, so I thought it would be worth mentioning.

### Further technical details

EF Core version: 3.1.1
Database provider: Microsoft.EntityFrameworkCore.InMemory 3.1.1, Microsoft.EntityFrameworkCore.Sqlite 3.1.1, Microsoft.EntityFrameworkCore.SqlServer 3.1.1
Target framework: .NET Core 3.1
Operating system: Windows 10
IDE: Visual Studio 2019 16.4.2

**Comment by ajcvickers** [comment]
Note for team: repros on master.

**Comment by maumar** [comment]
We should be using left join here instead

**Comment by Alirezanet** [comment]
I think i have same problem reported [here](https://github.com/dotnet/efcore/issues/19764) 
so +1

**Comment by maumar** [comment]
discrepancy comes from the fact that when navigation rewrite processes Count, we have optimization in place which just applies count on the source directly, completely ignoring pending selector. Assumption is that the count *should* be the same.

However, if the projection contains required navigation, and the entity that is on the other side of that navigation is filtered out by query filter, we remove the source to maintain referential integrity.

**Comment by maumar** [comment]
Filed #19801 to introduce model validation for this case as it seems counter-intuitive and can lead to issues.

**Comment by maumar** [comment]
per design discussion, added a model validation phase that warns against using required navigations where required side has the query filter and the optional side doesn't (tracked by #19801). Also improved docs to illustrate the issue, closing.

**Comment by hakanaltindis** [comment]
I have same problem too. I get this problem with these versions:
Target Framework: Net6.0
OS: Windows 10
Microsoft.EntityFrameworkCore.Relational: 6.0.8
Npgsql.EntityFrameworkCore.PostgreSQL: 6.0.6

Is there any roadmap to solve this issue?

**Comment by roji** [comment]
@hakanaltindis please open a new issue with a minimal, runnable code sample that shows the problem occurring.

**Comment by hakanaltindis** [comment]
> @hakanaltindis please open a new issue with a minimal, runnable code sample that shows the problem occurring.

I opened [new issue](https://github.com/dotnet/efcore/issues/29599)

**Comment by hakanaltindis** [comment]
Hi @roji ,

I read this topic before. And it says you may encounter unexpected results because global query filter is included in automatically. But I am aware of it and my all results come my expected.

The point what I do not get that Why does EF core not add the global query filters to get number of records.

**Comment by hakanaltindis** [comment]
When I call this LINQ `_dbContext.Activities.Include(a => a.Customer).Count();`, it generates this sql query `Select Count(*) From Activity`.

But when I call this LINQ `_dbContext.Activities.Include(a => a.Customer).ToList();`, please read carefully, it ends with ToList(). After that it generates this query
```sql
Select a.id, a.Name, c.Name
From Activity as a
	Inner Join Customer As c On (a.CustomerId = c.Id)
Where c.IsDeleted = false
```

What I try to say that If I want to get records/rows, everything works perfect. But If I just want to get number of records, it is not working correctly.

**Comment by roji** [comment]
@hakanaltindis that behavior - and the discrepancy between Count() and ToList() - is explained [above](https://github.com/dotnet/efcore/issues/19649#issuecomment-582154313). I agree it's somewhat odd, but that's why we have the warning I referred to above.

**Comment by hakanaltindis** [comment]
@roji  yes, it is odd. So do you have any plan to fix it? Because this odd prevents consistency.

**Comment by roji** [comment]
> I understand we can not use EF Core for Enterprise applications from what you say.

That is definitely not what I'm saying.

* When EF sees `_dbContext.Activities.Include(a => a.Customer).Count()`, it ignores (removes) the Include, since including the related entities generally isn't necessary for the counting. The general meaning of Include is to "bring back the related entities", which makes no sense in the context of a Count operation.
* However, if Customer is a required navigation, this has the side-effect of not applying any query filter on Customer, and so you get all Activities. Again, remember that Include is about loading related entities, not filtering.
* If we stopped ignoring the Include above, other users may start complaining since they're relying on the Count **not** filtering (and consider EF "not usable for Enterprise applications", as you wrote).

In short, although I agree the behavior is odd, there doesn't seem to be an obvious correct behavior here, and some users would complain no matter what we do. This is why the warning (and documentation) exists, to alert people to this behavior.

Note that you can always explicitly ask to filter as you wish by adding a Where clause before the filter (instead of the Include) - that should give you the results you want.

**Comment by hakanaltindis** [comment]
I am really trying to understand you. Sometimes your thoughts about `include()` comes to me true. I indicated before. I am really confused on the below example.

```C#
// Numbers are related with my example.
_dbContext.Activities.Include(a => a.Customer).Count(); // returns 11
_dbContext.Activities.Include(a => a.Customer).ToList().Count(); // returns 9
```

But you specified how odd this behavior.

And I don't want to prolong this topic. So be it. :)

I try to find a workaround for my case.

Thank you for your responses.

**Comment by hakanaltindis** [comment]
Thank you, @roji . I will follow this topic very closely. 👍 🙂

**Comment by smartouss** [comment]
Hello,

I'm facing the same issue and i see this have been reported from long ago.

This is a very weird behavior since its expected to when you call a Count() on a query to get the correct number. However i see there is a consideration for existing system so maybe a new method can be introduced to perform a Count with the include filter.

This is a killer problem if someone didn't notice this from the beginning like my case as i need to go and refactor all queries in my system that have such case.

**Comment by roji** [comment]
@smartouss I can understand the behavior may seem odd, but it's important to understand that Include() semantically **only** means that a related entity is to be included; any filtering behavior (i.e. the INNER JOIN being added for required navigations) is ""accidental"": EF could have chosen to implement the Include() via LEFT JOIN (see #29645), in which case Include() wouldn't cause any filtering behavior.

As written above, if you're looking to filter, then do just that - add a Where() clause that applies whatever filtering behavior you want, rather than relying on Include() to do so. I definitely don't see us introducing a special new method just for Count() that behaves differently here with regards to Include(), as you already have all the LINQ operators you need here.

**Comment by smartouss** [comment]
@roji I understand that the way if the implementation might be different but currently it's an Inner Join which do filter if a record is not there,  so technically Include is actually filtering the end result if ToList is called.

Regarding your suggestion of using Where, this not practical at all as it breaks the whole idea of defining a Global Query Filter, what the point of defining one if I'm going to apply it in most of my listing calls that require count or paging?

**Comment by roji** [comment]
> so technically Include is actually filtering the end result if ToList is called

Yep, Include() uses INNER JOIN which filters. But when you use Include() in LINQ, what you're doing is Including(), not filtering. If you want to filter, use Where. If you have a query that does Include() with an intent to filter, and then Count() (which doesn't care about included related entities), then you're doing it wrong - you need to replace the Include() with Where().

> Regarding your suggestion of using Where, this not practical at all as it breaks the whole idea of defining a Global Query Filter, what the point of defining one if I'm going to apply it in most of my listing calls that require count or paging?

I don't understand, what do global query filters have to do with Include() and whether it filters or not?

**Comment by smartouss** [comment]
I'm not using Include in intent to filter, what happen is I register a QueryFilter and this is filtering the result when calling Include, So the intent is including the result but EF is designed to apply the QueryFilter for that include in the query, but ignore it in the Count, this is causing weird behavior which in my opinion is incorrect, since the Count should return the number of items that query will return.

Regarding the relation between Include and GlobalFilter. The problem as stated above, when a filter registered to an entity and that entity is used in Include, this filter is applied, and you suggested to re apply the filter for that entity in any place we call the Count to get the correct result we expect. Now this causing a lot of confusion, It should either include in all places or ignore or provide a way to include it in specific cases, because your suggestion means i have to update the code in more than 50+ place for example.

**Comment by roji** [comment]
If I'm understanding the above correctly, you're using Include() on a query which then does Count(), is that correct? If so, why are you doing that, given that Include() is about including related entities in entities you're fetching, but you're not fetching any entities at all since you're calling Count()?

Otherwise I'm still a bit confused about what you're describing. Can you please include a brief code sample so it's clearer what you're doing here?

**Comment by smartouss** [comment]
My case is actually just needing to include the data of some entities, but the problem is i need paging and for that i need to perform count to get the total number of items, that's why im performing count.

So im building my query, doing count, then performing the actual query to return result, following is an example:
```C#
var query = dbcontext.EntityOne.Inculde(e => e.EntityTwo).Where(e => e.name.Contains("Value"));
var result = new PagedResult<EntityOne>();
result.TotalCount = query.Count();
result.Result = await query.Skip(pageSize * pageNumber).Take(pageSize).ToListAsync();
return result;
``` 

So as you can see my intention is just to include the entity two, but its filterting my result since it have a Global Query Filter, and im getting count 5 while im getting only 3 items. And this happening in all my listing apis and some have multiple includes, so applying a custom filter everywhere for this case is not practical and not efficient.

**Comment by roji** [comment]
Thanks for providing the extra context here, that makes sense. This is basically the exact scenario called out [in our docs](https://learn.microsoft.com/en-us/ef/core/querying/filters?tabs=ef10#query-filters-and-required-navigations) when combining required navigations and global query filters; read that section for the possible ways to address this (e.g. define the navigation as optional, causing EF to generate a LEFT JOIN instead). As written above, #29645 could be a change we'd do in EF to mitigate this.

Aside from this, I advise reading [our documentation on pagination](https://learn.microsoft.com/en-us/ef/core/querying/pagination) to understand why Skip and Take could be a problematic way to implement pagination.

**Comment by smartouss** [comment]
Thanks for providing the needed documentations, however this still won't help us, as we require pagination based on page number and as per the documentation the keyset pagination does not support it.

Another thing i will add to the context of the issue is why we using the global filter, we have soft delete for all our entities in the system, and we are using the global filter to automatically filter them out, so still the behavior in the returned result that i want is kinda correct, so the solution for me is to do cascade soft delete for now.

I would again suggest to have a way to force the global filter in the count operation as it will make the life of developers easier for such cases and make EF more easier to use in such cases.

**Comment by roji** [comment]
> I would again suggest to have a way to force the global filter in the count operation as it will make the life of developers easier for such cases and make EF more easier to use in such cases.

This is something that you're suggesting as a very specific fix to your very specific problem, but which I believe doesn't make sense in the more global context of EF. For example, there are other (possibly theoretic) cases where EF could be eliding Includes, since the included entities aren't needed for the final query. For example, there could be a (possibly auto-included) Include somewhere, but then a Select which projects out a property that's not on the included entity; in this case, the Include can (and should) be elided, but doing so would again affect query results if the Include otherwise generates an INNER JOIN which has the filtering effect.

As I wrote above, the proper fix here would be #29645, i.e. make Include itself completely side-effect by changing it to generate LEFT JOIN instead of INNER JOIN. Another possible direction would be to stop eliding Includes altogether when they're not necessary (as we're doing for Count), though that could be bad for perf as JOINs are left in the query when they're not needed (especially when auto-includes are defined).

I again recommend looking at configuring your included navigation as optional - this would cause EF to generate LEFT JOIN instead of INNER JOIN today, and cause this problem to go away.
